packages/@lwc/engine-core/src/framework/utils.ts

```diff
@@ -98,3 +100,18 @@ export function flattenStylesheets(stylesheets: TemplateStylesheetFactories): St
     }
     return list;
 }
+
+// Set a ref (lwc:ref) on a VM, from a template API
+export function setRefVNode(vm: VM, ref: string, vnode: VBaseElement) {
+    // Create the refVNodes object on-demand. This ensures that templates with no refs
+    // don't pay the perf tax of creating objects unnecessarily.
+    const refVNodes: RefVNodes = isNull(vm.refVNodes)
+        ? (vm.refVNodes = create(null))
+        : vm.refVNodes;
+
+    // In cases of conflict (two elements with the same ref), prefer, the last one,
+    // in depth-first traversal order.
+    if (!(ref in refVNodes) || refVNodes[ref].key < vnode.key) {
```

Comment:
Unlike Vue which is setting the refs during patching, this change set the ref during the template evaluation. 

With this approach, there is a window where the component can access the `refs` property but when the elements aren't created yet.

```html
<template>
  <div lwc:ref="first">first</div>
  <div data-foo={foo}>second</div>
</template>
```

In the example above, if the `foo` getter accesses `this.refs.first` it would return `undefined`. I don't think it's a big deal, but it's something worth calling out.

Reply:
If you think it's important, we can do it more like Vue. What would be the equivalent point in our code where we'd set the refs?

Reply:
I would say probably in [`patch`](https://github.com/salesforce/lwc/blob/04d8201eecd3e6810506072db999381ee86066bc/packages/@lwc/engine-core/src/framework/rendering.ts#L79) and [`mount`](https://github.com/salesforce/lwc/blob/04d8201eecd3e6810506072db999381ee86066bc/packages/@lwc/engine-core/src/framework/rendering.ts#L124) in `rendering.ts`.

Reply:
Hm, to get that to work, though, we'd have to pass the `RefVNodes` object into `mount` and `unmount`, correct? Seems a bit awkward for the function signature, but I agree it's elegant. I'll take a look.

Reply:
At some point, we would have to pass the `VM` to those rendering functions (eg. for scoped slots). My recommendation would be to pass the `VM` instead of the `RefVNode`.

Reply:
@pmdartus I [experimented](https://github.com/nolanlawson/lwc-1/commit/8d20284bfa23742729ab7dd17c35b1fa19364809) with this, but it actually breaks this test (`this.refs.coinflip` is undefined):

https://github.com/salesforce/lwc/blob/2ffa2dba6b4cc8852484c2d64a6d806e912aa678/packages/%40lwc/integration-karma/test/component/refs/index.spec.js#L210

The reason is that in this template:

https://github.com/salesforce/lwc/blob/2ffa2dba6b4cc8852484c2d64a6d806e912aa678/packages/%40lwc/integration-karma/test/component/refs/x/conditional/conditional.html#L1-L10

... we mount the `coinflip` element inside of the `if:true`, then unmount the one in the `if:false`. So we end up updating `coinflip` and then removing it from the `refs` object.

There is maybe a clever way of making this work (e.g. doing all the unmounting before we do any of the mounting), but I'm not sure it's worth it?

That said, I am concerned about your example (`<div data-foo={foo}>second</div>`), and I'm wondering if the current implementation is truly resilient.

Reply:
Good catch. I have done some extra digging in Vue's internals: [playground](https://stackblitz.com/edit/vue-upe5q1?file=src/components/HelloWorld.vue).

The way Vue approaches this is by applying the refs after the DOM is patched and before invoking the updated life-cycle callback. During patching, all the ref settings are queued: [`doSet`](https://github.com/vuejs/core/blob/fbd697a4b6c238147d661026836314862fd09221/packages/runtime-core/src/rendererTemplateRef.ts#L85). This queue is then flushed after the patching process: [`flushPostFlushCbs`](https://github.com/vuejs/core/blob/fbd697a4b6c238147d661026836314862fd09221/packages/runtime-core/src/scheduler.ts#L157) (async in the case of Vue).

I think Vue piggybacks on this concept of post flush queue already in use for [lifecycle hooks invocations](https://github.com/vuejs/core/search?l=TypeScript&q=queuePostRenderEffect), which is currently foreign to LWC.  

I am also unsure if it's the correct approach or not given the amount of refactoring it would require us to implement. I would suggest keeping the current approach for now and address it in the future if needed. Fixing this issue shouldn't cause much issue because I don't expect components to access their refs in the middle of the rendering. 

As I am writing this, I am now wondering if a good way to mitigate this is to prevent developers from accessing the refs in the middle of the rendering cycle since the DOM isn't stable.

Reply:
> As I am writing this, I am now wondering if a good way to mitigate this is to prevent developers from accessing the refs in the middle of the rendering cycle since the DOM isn't stable.

This seems doable. I'll look into it.

Reply:
@pmdartus I added a condition that returns `undefined` for `this.refs` if the template is currently being rendered: 8ccb9d9c5

What's interesting about this situation is that we actually _have_ to return `undefined`, because otherwise we cache the `refs` based on the `refVNodes` and it can become permanently out-of-date. I added a test to verify this condition.

Really weird situation, but hopefully it won't actually happen very frequently!